src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -620,24 +621,9 @@ public static unsafe bool TryCompleteReceiveFrom(SafeCloseSocket socket, byte[]
                 {
                     // Special case a receive of 0 bytes into a single buffer.  A common pattern is to ReceiveAsync 0 bytes in order
                     // to be asynchronously notified when data is available, without needing to dedicate a buffer.  Some platforms (e.g. macOS),
-                    // however, special-case a receive of 0 to always succeed immediately even if data isn't available.  As such, we treat 0
-                    // specially, checking whether any bytes are available rather than doing an actual receive.
-                    receivedFlags = SocketFlags.None;
-                    received = -1;
-
-                    int available = 0;
-                    errno = Interop.Sys.GetBytesAvailable(socket, &available);
-                    if (errno == Interop.Error.SUCCESS)
-                    {
-                        if (available > 0)
-                        {
-                            bytesReceived = 0;
-                            errorCode = SocketError.Success;
-                            return true;
-                        }
-
-                        errno = Interop.Error.EAGAIN; // simulate a receive with no data available
-                    }
+                    // however return 0 instead of EAGAIN when data isn't available. As such, we treat 0 specially, and perform a 1-byte peek.
```

Comment:
I would change this from "however return 0 instead of EAGAIN when data isn't available" to something like "however complete a 0-byte read successfully when data isn't available, as the request can logically be satisfied synchronously".